readthedocs/api/v3/views.py

```diff
@@ -381,6 +384,32 @@ def create(self, request, **kwargs):  # pylint: disable=arguments-differ
         return Response(data=data, status=code)
 
 
+class NotificationsViewSet(
+    APIv3Settings,
+    NestedViewSetMixin,
+    ProjectQuerySetMixin,
+    FlexFieldsMixin,
+    ReadOnlyModelViewSet,
+):
+    model = Notification
+    lookup_field = "pk"
+    lookup_url_kwarg = "notification_pk"
+    serializer_class = NotificationSerializer
+    queryset = Notification.objects.all()
+    # filterset_class = BuildFilter
+
+    # http://chibisov.github.io/drf-extensions/docs/#usage-with-generic-relations
+    def get_queryset(self):
+        queryset = self.queryset.filter(
+            attached_to_content_type=ContentType.objects.get_for_model(Build)
+        )
+
+        # TODO: make sure if this particular filter should be applied here or somewhere else.
+        return queryset.filter(
+            attached_to_id=self.kwargs.get("parent_lookup_build__id")
+        )
```

Comment:
I think we can be flexible here. We can have 1) one API endpoint per resource or 2) only one global endpoint with a filter per resource.

I started doing 1) as a test because is what DRF extensions allows to do and it's a pattern we have been following for other endpoints too. Also, I think it's easier to understand what notifications you are getting back. An example of the request would be: 

* Notifications for a Build: `/api/v3/project/docs/builds/123456/notifications/`
* Notifications for a Project: `/api/v3/project/docs/notifications/`
* etc..

The particular view where you commented (`NotificationsViewSet`) would cover 2) --it's not finished, and can probably be commented out/removed for now-- and would require the user sending the `?attached_to=build` and also `?attached_to_id=123456` so the endpoint knows what's the exact object to retrieve notifications for: `/api/v3/notifications/?attached_to=build&attached_to_id=123456`

I didn't finish this because I wanted to discuss a little more with you before making the final decision. I personally prefer 1) since it's a cleaner UX and it's already working 😄 . Do you have any preference here?

However, the view `NotificationsViewSet` may also be required for a different usage: _mark a specific notification object as read_ via the URL: `PATCH /api/v3/notifications/718273/`. So, we may keep it for that purpose.

We don't have to decide this _right now_, since it's outside the scope of this PR, tho.

Reply:
I opened an issue for this at https://github.com/readthedocs/readthedocs.org/issues/10984

Reply:
I'll follow along that issue, but to note here:

>  I personally prefer 1) since it's a cleaner UX and it's already working

I think I prefer this too. It doesn't make much difference to the front end as we need to resolve the API URL and pass that into the JS view either way.

> However, the view NotificationsViewSet may also be required for a different usage

Good point. On the front end, the API should give a URL for dismissing the notification in the API response, so the JS doesn't have to reconstruct this. So either way works here too, and it wouldn't be any more/less work to PATCH a deeper URL like `/api/v3/project/docs/builds/123456/notifications/789`